File: Programmers/DFS_BFS/travel_route.py
answer = []
def DFS(tickets, idx, visited, curr_list):
    v = visited[:]
    curr_list.append(tickets[idx][1])
    v[idx] = True
    last_city = curr_list[-1]

    if v.count(False) == 0:         #모든 경우를 다 살펴봤으니 answer에 추가하고 return
        answer.append(curr_list[:])
        curr_list.pop()             #현재까지 경로 중 이번에 추가한 경로 삭제
        return              

    for i in range(len(tickets)):   #i번째 ticket의 출발지가 현재 들린 마지막 도시이면 DFS 진행
        if v[i] is True:
            continue
        if last_city == tickets[i][0]:
            DFS(tickets, i, v, curr_list)

    curr_list.pop()


def solution(tickets):
    for i in range(len(tickets)):
        visited = [False for _ in range(len(tickets))]
        if (tickets[i][0] == "ICN"):
            curr_list = ["ICN"]
            DFS(tickets, i, visited, curr_list) #첫번째 도시가 ICN이면 DFS 진행

    answer.sort()       #answer를 알파벳 순으로 정렬
    print(answer)
    return answer[0]    #가장 첫번째 정답 출력

File: Programmers/DFS_BFS/travel_route.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<string> answer;

bool compare(vector<string> a, vector<string> b){
    return a[1] > b[1];
}

void DFS(int curr, vector<string> &route, const vector<vector<string>> &tickets, vector<bool> &used){    
    if(curr == tickets.size()) {
        answer = route;
        return;
    }
    
    for(int i= 0; i < tickets.size(); i++){
        if(tickets[i][0] == route[curr] && !used[i]){
            route.push_back(tickets[i][1]);
            used[i] = true;
            DFS(curr + 1, route, tickets, used);
            used[i] = false;
            route.pop_back();
        }
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end(), compare);
    vector<bool> used(tickets.size(), false);
    vector<string> route = {"ICN"};
    
    DFS(0, route, tickets, used);
    
    return answer;
}
